Add an e2e test to verify that pods are deleted on nodes

yujuhong opened the pull request:
The test verifies that kubelet deletes the pods/containers within a
reasonable time. It queries the kubelet /runningpods endpoint to retrieve a
list of running pods directly. The outline of the test is:
- Create an RC
- Wait until all pods are running (based on the pod status)
  status)
- Verify pods are running by querying the /runningpods
- Delete the RC
- Check all pods are deleted by querying /runningpods

yujuhong committed:
Add a kubelet /runningpods endpoint

/runningpods returns a list of pods currently running on the kubelet. The list
is composed by examining the container runtime, and may be different from the
desired pods to run known by kubelet.

This is useful for tests to verify that pods are indeed deleted on nodes.

yujuhong committed:
Add an e2e test to verify that pods are deleted on nodes

The test verifies that kubelet deletes the pods/containers within a reasonable
time. It queries the kubelet /runningpods endpoint to retrieve a list of
running pods directly. The outline of the test is:

 - Create an RC
 - Wait until all pods are running (based on the pod status)
 - Verify pods are running by querying the /runningpods
 - Delete the RC
 - Check all pods are deleted by querying /runningpods

yujuhong commented:
Risk: low. This PR adds an endpoint in the kubelet server and use it for the new e2e test. The endpoint is not used by any other production code, so it should be pretty safe.
Value: Currently we don't have any test verifying that kubelet actually deletes the pods/containers on the node. This new test would be able to verify that, and also signal if deletion took an abnormally long time.

yujuhong commented:
/cc @dchen1107

JasonGiedymin commented on test/e2e/kubelet_stats.go:
```diff
@@ -159,17 +160,32 @@ func HighLatencyKubeletOperations(c *client.Client, threshold time.Duration, nod
-// Retrieve metrics from the kubelet server of the given node.
-func getKubeletMetrics(c *client.Client, node string) (string, error) {
-	metric, err := c.Get().
+// Performs a get on a node proxy endpoing given the nodename and rest client.
```
Spelling of word endpoint.

yujuhong replied:
Fixed the typo. Thanks!

yifan-gu commented on pkg/kubelet/kubelet.go:
```diff
@@ -1834,6 +1834,24 @@ func (kl *Kubelet) GetPods() []*api.Pod {
 
+// GetRunningPods returns all pods running on kubelet from looking at the
+// container runtime cache. This function converts kubecontainer.Pod to
+// api.Pod, so only the fields that exist in kubecontainer.Pod are considered
```
Do you really mean `kubecontainer.Pod`?

yujuhong replied:
Modified it to make it clear that only common fields (of kubecontainer.Pod and api.Pod) are meaningful.

yujuhong commented:
The new push moves /runningpods handler registration to `InstallDebuggingHandlers` since it's only used for testing.

quinton-hoole commented on pkg/kubelet/container/runtime.go:
```diff
@@ -309,3 +309,21 @@ func ParsePodFullName(podFullName string) (string, string, error) {
+
+// ToAPIPod converts Pod to api.Pod. Note that if a field in api.Pod has no
+// corresponding field in Pod, the field would not be populated.
+func ToAPIPod(pod *Pod) *api.Pod {
```
nit: Doesn't it make more sense to have this be a member of Pod?  I don't feel terribly strongly about this one.

quinton-hoole commented on pkg/kubelet/server.go:
```diff
@@ -296,6 +299,28 @@ func (s *Server) handlePods(w http.ResponseWriter, req *http.Request) {
+// handleRunningPods returns a list of pods running on Kubelet. The list is
+// provided by the container runtime, and is different from the list returned
+// by handlePods, which is a set of desired pods to run.
+func (s *Server) handleRunningPods(w http.ResponseWriter, req *http.Request) {
```
This is mostly a cut 'n paste of the above function.  Please factor out the common code.

quinton-hoole commented on pkg/kubelet/kubelet.go:
```diff
@@ -1834,6 +1834,23 @@ func (kl *Kubelet) GetPods() []*api.Pod {
+// container runtime cache. This function converts kubecontainer.Pod to
+// api.Pod, so only the fields that exist in both kubecontainer.Pod and
+// api.Pod are considered meaningful.
+func (kl *Kubelet) GetRunningPods() ([]*api.Pod, error) {
```
The kubelet code all looks reasonably ok to me barring my comments, but cc @dchen1107 for a second set of eyes, as she is much more familiar with this code than I am.

quinton-hoole commented on test/e2e/kubelet_deletes.go:
```diff
@@ -0,0 +1,146 @@
+/*
```
I wonder whether we don't rather want a test/e2e/kubelet.go to contain all kubelet tests?  Most of the other components follow this convention.

yujuhong replied:
It depends. There are also "service.go" and "service_latency.go". 

I think kubelet is a big enough component that grouping all the tests in one file may be hard to read in the future. Given that this is easy to change and we don't have other tests yet, I think it might be okay to keep the filename. I can change if you insist.

quinton-hoole replied:
Yes, I'd rather rename the file to kubelet.go, even if you don't merge it with kubelet_stats.go.

quinton-hoole commented on test/e2e/kubelet_deletes.go:
```diff
@@ -0,0 +1,146 @@
+	}
+
+	deleteTests := []DeleteTest{
+		{podsPerNode: 10, timeout: 3 * time.Minute},
```
Does it really take as long as 3 minutes to delete 10 pods?  Can we shorten this?

yujuhong replied:
No, at least not on my GCE cluster. However, this is more of a correctness than a performance test. The timeout is long enough that i'd not fail randomly due to other issues, but it will still surface any kind of hang in kubelet. Plus, with a slightly longer timeout, we can actually print out the latency in the end (without failing prematurely).

quinton-hoole replied:
See below - we really need to get the maximum fail time to about 2 minutes, somehow.  I think it's probably around 8 minutes currently.

quinton-hoole replied:
To be clear, my concern is not about performance vs correctness.  It's a far more pedestrian concern about how long it takes for the tests to run, both manually and in continuous integration systems like Jenkins.  We have about 100 e2e tests now, and all of those minutes add up.

quinton-hoole commented on test/e2e/kubelet_deletes.go:
```diff
@@ -0,0 +1,146 @@
+			totalPods := itArg.podsPerNode * numNodes
+
+			By(fmt.Sprintf("Creating a RC of %d pods and wait until all pods of this RC are running", totalPods))
+			rcName := "cleanup" + strconv.Itoa(totalPods) + "-" + string(util.NewUUID())
```
nit: You don't strictly speaking need the UUID here, as you're running this in your own namespace, and totalPods is already unique within this test invocation.

quinton-hoole replied:
And another nit - isn't it simpler just to use a single Sprintf("cleanup%d-%s",...) to initialize rcName?

yujuhong replied:
1. UUID seems safe if we ever expand the parameters of the test, but I'm happy to remove it.
2. I was following the naming code pattern that used in most (if not all) e2e tests. I'll switch to Sprintf

yujuhong replied:
I am keeping the UUID just to be safe now that we are going to check the pod name prefix (which is the RCName).

quinton-hoole commented on test/e2e/kubelet_deletes.go:
```diff
@@ -0,0 +1,146 @@
+			// Perform a sanity check so that we know all desired pods are
+			// running on the nodes according to kubelet.
+			seen := util.NewStringSet()
+			for _, name := range nodeNames.List() {
```
In theory some pods might have failed, to be restarted on other nodes.

yujuhong replied:
Yes, In theory this is possible, but I think this is not likely to happen in a controlled test. RunRC waits until all pods are running before returns, so for this to happen, some pod needs to fail on its own or gets OOM killed.  

Since we are using "non-existence" of the pods from (/runningPods) to determine successful deletions, the sanity check here ensures "/runningPods" and our checking code works correctly.

What I can do to make this more robust are:
 1. Don't construct a list of nodes with pods running, and simply use all nodes (so the list doesn't become stale).
 2. Poll until all pods are running.

What do you think?

quinton-hoole replied:
Yes, that's fine.  See more comments below on the topic.
Just to be clear, for 2. I assume that you mean "poll until all of the pods created in this test are NOT running".  Don't forget that your test might be running in parallel with many other tests (even though they should be running in different namespaces).

quinton-hoole commented on test/e2e/kubelet_deletes.go:
```diff
@@ -0,0 +1,146 @@
+					seen.Insert(string(pod.UID))
+				}
+			}
+			Expect(seen.IsSuperset(desired)).To(BeTrue(), "All pods should be running on nodes")
```
See above.  I don't think that this is strictly speaking true.  Pods could have failed and been moved. At the very least please have the text description reflect that this is perhaps unusual but not theoretically incorrect, in case someone needs to debug this in future.

yujuhong replied:
See the comment above.

quinton-hoole commented on test/e2e/kubelet_deletes.go:
```diff
@@ -0,0 +1,146 @@
+			//   - a bug in graceful termination (if it is enabled)
+			//   - docker slow to delete pods (or resource problems causing slowness)
+			start := time.Now()
+			// TODO: parallellize the checking by nodes if the number of nodes
```
I'd strongly suggest that you do this up front, in this PR.  
As it currently stands, even on a 2-node test cluster, this test if going to take > 6 minutes just to timeout checking the two nodes, and even longer to actually fail, as far as I can tell.  You ideally want to get that under 1 minute, or 2 minute if it's absolutely necessary.

yujuhong replied:
Sure, I will look into parallelizing it.

FWIW, I think it's safe to change the timeout to 1 minute, given that all my runs so far finished < 6 seconds. To put things into perspective, the current podStartTimeout is 5 minutes.

quinton-hoole replied:
If they finish in 6 seconds, I'd set the timeout to maybe 30 seconds.  Your target should be around the 99th percentile.

quinton-hoole commented on test/e2e/kubelet_deletes.go:
```diff
@@ -0,0 +1,146 @@
+			// is significant (e.g. >10).
+			for _, name := range nodeNames.List() {
+				Logf("Checking pods on node %v via /runningpods endpoint", name)
+				Expect(waitTillKubeletDeletesPods(framework.Client, desired, name, itArg.timeout)).NotTo(HaveOccurred())
```
Similar argument to above regarding pods or nodes that might have failed and been rescheduled during the running of these tests.  I think that you're going to need to label the pods as being associated with the RC, and then repeatedly check across all nodes in the cluster that pods with that label no longer exist.

yujuhong replied:
I see your point, but this is going to be tricky. docker containers don't keep the original label information around. I can change this, but I am not sure it's a good time to add this purely for the tests. I could do the next best thing and check the pod names (which should be prefixed by the RC name), and also scan all the nodes. Does that sound okay to you?

quinton-hoole replied:
Yes, that sounds fine, using the pod name prefix as a proxy for a label.  Just comment the code appropriately to explain that it's a bit of a hack to get around the docker limitation.

quinton-hoole commented on test/e2e/kubelet_stats.go:
```diff
@@ -159,17 +160,32 @@ func HighLatencyKubeletOperations(c *client.Client, threshold time.Duration, nod
-func getKubeletMetrics(c *client.Client, node string) (string, error) {
-	metric, err := c.Get().
+// Performs a get on a node proxy endpoint given the nodename and rest client.
+func nodeProxyRequest(c *client.Client, node, endpoint string) client.Result {
```
Nice factorization - thanks for this.

wojtek-t commented on pkg/kubelet/server.go:
```diff
@@ -148,6 +149,8 @@ func (s *Server) InstallDebuggingHandlers() {
 	s.mux.HandleFunc("/containerLogs/", s.handleContainerLogs)
 	s.mux.Handle("/metrics", prometheus.Handler())
+	// The /runningpods endpoint is used for testing only.
+	s.mux.HandleFunc("/runningpods", s.handleRunningPods)
```
nit: maybe "runningPods" instead (i.e. "P" instead of "p" in pods)

quinton-hoole replied:
I think that one could argue this one either way.  I guess that we should at least be consistent one way or the other, which we're not currently.  See "syncloop" vs "containerLogs" and "portForward".

yujuhong replied:
Hmm...I missed this comment. I could do a follow-up PR to make the names consistent, whichever format we decide to use. What do you think?

quinton-hoole replied:
Yes, that's totally fine.

quinton-hoole commented:
OK, seems like we know what to do to finish this one off.  Re-assigning back to @yujuhong to make the final tweaks.  Please assign back to me when it's ready for final review.

quinton-hoole commented:
e2e failure looks like a flake.  Trying again.
@k8s-bot test this please

quinton-hoole commented:
@k8s-bot test this please

quinton-hoole commented:
ping @yujuhong .  Any progress?  Seems we're very close to merging this one.

yujuhong commented:
@quinton-hoole, sorry for the delay, I was OOO on Friday. I am working on the PR and will upload soon.

yujuhong commented:
Rebased and addressed the comments (shorten the timeout, parallelizing the pod checking, switching to use pod name prefix to filter pods). PTAL!

Note that the test assumes that the pods created by RC will be spread somewhat evenly among the nodes, but this does not happen because of #10242 (for pods in the non-default namespace). As the results, all pods may be scheduled onto one node until the issue is resolved. The test will still pass, but only one node is actually tested.

yujuhong commented:
@k8s-bot ok to test.

quinton-hoole commented:
LGTM from me, but I'd really like one more pair of eyes on the kubelet code before we merge this.  Assigning to @dchen1107 for LGTM, ok-to-merge, although @lavalamp could also do this if he has time.

dchen1107 commented:
I do have concern to introduce arbitrary endpoint to kubelet, but since you only register this for debugging handler, which should be ok. 

LGTM. I only reviewed kubelet changes, and totally trust @quinton-hoole 's review on testing side. :-)

yujuhong commented:
go v1.3 doesn't like the `for range {` syntax. Fixed that and pushed again. The rest of the code remains unchanged.

lavalamp commented on pkg/kubelet/container/runtime.go:
```diff
@@ -284,6 +284,24 @@ func (p *Pod) FindContainerByName(containerName string) *Container {
+// corresponding field in Pod, the field would not be populated.
+func (p *Pod) ToAPIPod() *api.Pod {
+	var pod api.Pod
+	pod.UID = p.ID
```
ugh, ideally we would keep api.Pod and kubelet/container.Pod close enough that pkg/conversion could do this automatically :/

Not asking for a change.

yujuhong replied:
In the current stage, they are far from close. I am not sure if the plan is to make them close, or to grow them differently. @dchen1107 would know more...

yujuhong commented:
> go v1.3 doesn't like the for range { syntax. Fixed that and pushed again. The rest of the code remains unchanged.

ok...I pushed again because gofmt (v1.4) didn't like that version. Hopefully this one would please both.
